src/audio/module_adapter/module/cadence.c

```diff
@@ -207,16 +254,100 @@ static int cadence_codec_post_init(struct processing_module *mod)
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_4
 static int cadence_codec_init(struct processing_module *mod)
 {
-	int ret;
+	const struct ipc4_cadence_module_cfg *cfg;
+	struct module_data *codec = &mod->priv;
+	struct cadence_codec_data *cd;
+	struct module_config *setup_cfg;
 	struct comp_dev *dev = mod->dev;
+	int ret;
+
+	comp_dbg(dev, "cadence_codec_init() start");
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(struct cadence_codec_data));
+	if (!cd) {
+		comp_err(dev, "cadence_codec_init(): failed to allocate memory for cadence codec data");
+		return -ENOMEM;
+	}
+
+	codec->private = cd;
+	codec->mpd.init_done = 0;
+	cd->self = NULL;
+	cd->mem_tabs = NULL;
+	cd->api = NULL;
+	cd->setup_cfg.avail = false;
```

Comment:
why did you use `rzalloc()` then and now you're initialising values to zeros again?